### PR TITLE
Various

### PR DESCRIPTION
**Pull request by sinkingsugar**

### CODE REVIEW

File: src/core/async.hpp [hunk 1]
```diff
@@ -20,79 +112,100 @@ struct SharedThreadPoolConcurrency {
...
 #if defined(__EMSCRIPTEN__) && !defined(__EMSCRIPTEN_PTHREADS__)
   return func();
 #else
-  SHVar res{};
-  std::packaged_task<SHVar()> task(std::forward<FUNC>(func));
-  auto future = task.get_future();
+  struct BlockingCall : TidePool::Work {
+    BlockingCall(FUNC &&func) : func(std::move(func)), exp(), res(), complete(false) {}
+
+    virtual ~BlockingCall() {}
```
Comment (guusw) [inline src/core/async.hpp:126]: Doesn't need virtual destructor, only if you were to destroy it from it's base class, but then you'd need it on Work instead

**Comment by guusw** [comment]
Just some style comments but looks good

File: src/core/async.hpp [hunk 2]
```diff
@@ -1,11 +1,103 @@
...
+struct TidePool {
+  struct Work {
+    virtual void call() = 0;
+  };
+
+  TidePool() {
+    _running = true;
+
+    // spawn workers first
+    for (size_t i = 0; i < NumWorkers; ++i) {
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/core/async.hpp:30]: not the _running though, as it can race if destroyed immediately

File: src/core/async.hpp (hunk 1 above)
Comment (sinkingsugar) [reply src/core/async.hpp:134]: I had a warning, that's why I added it in test_runtime which is similar.
Doesn't hurt anyway does it?

Comment (sinkingsugar) [reply src/core/async.hpp:134]: no warning in this header though seems indeed

Comment (sinkingsugar) [reply src/core/async.hpp:134]: err at least on win, will remove for now we see

**Pull request merged**